edit/candidate.go

```diff
@@ -101,39 +108,27 @@ func outputComplexCandidate(ec *eval.EvalCtx,
 	ec.OutputChan() <- c
 }
 
-func (ed *Editor) filterAndCookCandidates(ev *eval.Evaler, completer string, pattern string,
-	cands []rawCandidate, q parse.PrimaryType) ([]*candidate, error) {
-
-	matcher, ok := ed.lookupMatcher(completer)
-	if !ok {
-		return nil, errMatcherMustBeFn
-	}
+func (ed *Editor) filterAndCookCandidates(ev *eval.Evaler, matcher eval.CallableValue, pattern string,
+	cands chan eval.Value, q parse.PrimaryType) ([]*candidate, error) {
 
-	input := make(chan eval.Value, len(cands))
 	ports := []*eval.Port{
-		{Chan: input}, {File: os.Stdout}, {File: os.Stderr}}
+		{Chan: cands}, {File: os.Stdout}, {File: os.Stderr}}
 	ec := eval.NewTopEvalCtx(ev, "[editor matcher]", "", ports)
 
-	// TODO(xiaq): Revise the API of completers so that the complter write
-	// directly to a channel and run concurrently with the matcher.
-	for _, cand := range cands {
-		input <- eval.String(cand.text())
-	}
-	close(input)
-
 	args := []eval.Value{eval.String(pattern)}
 	values, err := ec.PCaptureOutput(matcher, args, eval.NoOpts)
 	if err != nil {
 		return nil, err
-	} else if len(values) != len(cands) {
-		return nil, errIncorrectNumOfResults
 	}
 
 	var filtered []*candidate
-	for i, value := range values {
-		if eval.ToBool(value) {
-			filtered = append(filtered, cands[i].cook(q))
+	for _, value := range values {
+		if rc, ok := value.(rawCandidate); ok {
+			filtered = append(filtered, rc.cook(q))
```

Comment:
Another thing -- sorry for not raising this earlier -- it's perhaps better to collect raw candidates, sort them by `text()` before cooking them, instead of sorting cooked candidates. Suppose that you have a directory containing files `a`, `x`, `a b`, `x y`, sorting cooked candidates will sort them by the quoted version, resulting in the ordering `'a b'`, `'x y'`, `a`, `x`; I think it's better to sort them by the unquoted version, namely `a`, `'a b'`, `x`, `'x y'`.